pkg/controller/computenodeopenstack/computenodeopenstack_controller.go

```diff
@@ -395,45 +486,101 @@ func updateNodesStatus(c client.Client, instance *computenodev1alpha1.ComputeNod
 				err = deleteBlockerPodFinalizer(c, node.Name, instance)
 			}
 		case "Ready":
+			log.Info(fmt.Sprintf("Previous status Ready: %s", node.Name))
 			newNode := computenodev1alpha1.Node{Name: node.Name, Status: nodeStatus}
 			nodesStatus = append(nodesStatus, newNode)
 
 			if nodeStatus == "Ready" {
 				readyWorkers += 1
 			} else if nodeStatus == "SchedulingDisabled" {
-				// trigger drain (new pod)
-				err = triggerNodeDrain(c, node.Name, instance)
+				// trigger node drain pre tasks (new pod)
+				err = triggerNodePreDrain(c, node.Name, instance, openstackClientAdmin, openstackClient)
 				if err != nil {
 					return err
 				}
 			}
 		case "SchedulingDisabled":
-			// if drain has finished (pod completed) remove blocker-pod finalizer
-			drained, err := isNodeDrained(c, node.Name, instance)
-			if err != nil {
+			log.Info(fmt.Sprintf("Previous status SchedulingDisabled: %s", node.Name))
+			// if pre draining has finished (job completed) remove blocker-pod finalizer
+			drainedPre, err := isNodeDrained(c, node.Name, node.Name+"-drain-job-pre", instance)
+			if err != nil && errors.IsNotFound(err) {
+				log.Error(err, fmt.Sprintf("Pre NodeDrain job errors.IsNotFound: %s", node.Name+"-drain-job-pre"))
+				return nil
+			} else if err != nil {
 				return err
 			}
-			if drained {
-				err = deleteBlockerPodFinalizer(c, node.Name, instance)
+
+			log.Info(fmt.Sprintf("Pre draining job: %s", drainedPre))
+			switch drainedPre {
+			case "succeeded":
+				// taint node that daemonset pods get stopped
+				// since we loop over the nodes in status update we need to
+				// 1) taint the node to have OSP pods stopped
+				// 2) run post drain job
+				// 3) remove blocker pod to get node finally removed
+				err = addToBeRemovedTaint(kclient, node)
 				if err != nil {
 					return err
 				}
 
-				for i, nodeToDelete := range instance.Spec.NodesToDelete {
-					if nodeToDelete.Name == node.Name {
-						if len(instance.Spec.NodesToDelete) == 1 {
-							instance.Spec.NodesToDelete = []computenodev1alpha1.NodeToDelete{}
-						} else {
-							instance.Spec.NodesToDelete = removeNode(instance.Spec.NodesToDelete, i)
-						}
-						err = c.Update(context.TODO(), instance)
-						if err != nil {
-							return err
+				// trigger node drain post tasks (new pod)
+				err = triggerNodePostDrain(c, node.Name, instance, openstackClientAdmin, openstackClient)
+				if err != nil {
+					return err
+				}
+
+				// if post draining has finished (job completed) remove the pod
+				drainedPost, err := isNodeDrained(c, node.Name, node.Name+"-drain-job-post", instance)
+				if err != nil && errors.IsNotFound(err) {
+					log.Error(err, fmt.Sprintf("Post NodeDrain job errors.IsNotFound: %s", node.Name+"-drain-job-post"))
+					return nil
```

Comment:
why nil? this is an error right? what actions are we expected to do if that pod is not there?

Reply:
do we assume "not found" == succeeded and we trigger the rest of the cleanup, or recreate the drain-job-post?

Reply:
good catch, yes we should return an error.